### ISSUE TITLE
Require C++11 for compilation?

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#1033]
It's currently possible to (mostly) compile parts of mlpack without C++11, but it can make life difficult.  For instance, I spent ~an hour today trying to compile a simple FFN that I'd worked on earlier:

```
$ g++ -O2 -o run_network -I/home/ryan/src/mlpack/build-nodebug/include/ -L/home/ryan/src/mlpack/build-nodebug/lib/ -lmlpack -larmadillo -lboost_program_options run_network.cpp
```

and this output about a billion lines of errors because boost::variant had an ambiguous overload.  I dug for a while, then I realized, I just have to specify `-std=c++11` and it's fine.

So, is it time for us to require a C++11 compiler and output a compilation error via `#error` if we don't detect one?  I think it would be a good idea, I just want to open this ticket and let it sit a while before actually making the change, to see if there are any objections.  I think we can at least reasonably expect that everyone will have a fully functional C++11 compiler available now since it is 2017.  (Though this will make it harder for me to try to compile mlpack on that old Solaris 7 box I have been playing with!)

**Comment by mentekid** [comment]
My opinion is that this makes sense, I was not even aware you could compile mlpack without C++11, but I think it is a reasonable requirement.

For the Solaris 7 people (and how many could that be), they can always take the extra step of removing the #error lines.

Would this happen in the mlpack headers, or in CMake?

**Comment by rcurtin** [comment]
This change would be in the mlpack headers.  Now that I think about it, CMake will refuse to build mlpack itself without a C++11 compiler.  The problem I encountered arose when I tried to use non-C++11 compilation while including some mlpack headers.

**Comment by mentekid** [comment]
I see. Makes sense to do it in the header then.

My opinion is the same - C++11 is pretty much standard.

**Comment by rcurtin** [comment]
When I dug into this further, I found that C++11 happens to already be required via the header file `<utility>`.  Instead, it turns out what I was encountering was this bug:

https://svn.boost.org/trac10/ticket/11285

So actually my compiler was already by default in C++14 mode, but this causes that Boost bug with Boost 1.58.  In 37be9f4, I added a macro check to `prereqs.hpp` that detects this situation (C++14 and Boost < 1.59) and issues an error.  Should help save some other people pain...
